runtime/memprof.c

```diff
@@ -156,1176 +1258,983 @@ static memprof_domain_t domain_create(caml_domain_state *caml_state)
+ * copy it onto the Caml heap later, when we're ready to call the
+ * allocation callback. */
+static value capture_callstack_no_GC(memprof_domain_t domain)
 {
-  value res;
-  intnat callstack_len =
-    caml_collect_current_callstack(&callstack_buffer, &callstack_buffer_len,
-                                   callstack_size, alloc_idx);
-  CAMLassert(local->suspended);
-  res = caml_alloc(callstack_len, 0);
-  memcpy(Op_val(res), callstack_buffer, sizeof(value) * callstack_len);
-  if (callstack_buffer_len > 256 && callstack_buffer_len > callstack_len * 8) {
-    caml_stat_free(callstack_buffer);
-    callstack_buffer = NULL;
-    callstack_buffer_len = 0;
+  value res = Atom(0); /* empty array. */
+  size_t frames =
+    caml_get_callstack(Callstack_size(domain->entries.config),
+                       &domain->callstack_buffer,
+                       &domain->callstack_buffer_len, -1);
+  if (frames) {
+    callstack_stash_t stash = caml_stat_alloc_noexc(sizeof(callstack_stash_s)
+                                                    + frames * sizeof(value));
+    if (stash) {
+      stash->frames = frames;
+      memcpy(stash->stack, domain->callstack_buffer, sizeof(value) * frames);
+      res = Val_ptr(stash);
+    }
   }
+
+  shrink_callstack_buffer(domain, frames);
   return res;
 }
 
-/**** Managing data structures for tracked blocks. ****/
-
-/* Reallocate the [ea] array if it is either too small or too
-   large.
-   [grow] is the number of free cells needed.
-   Returns 1 if reallocation succeeded --[ea->alloc_len] is at
-   least [ea->len+grow]--, and 0 otherwise. */
-static int realloc_entries(struct entry_array* ea, uintnat grow)
+/* Capture the call stack when sampling an allocation from Caml. We
+ * have to deal with combined allocations (Comballocs), but can
+ * allocate the resulting call stack directly on the Caml heap. Should
+ * be called with [domain->current->suspended] set, as it allocates.
+ * May cause a GC. */
+static value capture_callstack_GC(memprof_domain_t domain, int alloc_idx)
 {
-  uintnat new_alloc_len, new_len = ea->len + grow;
-  struct tracked* new_t;
-  if (new_len <= ea->alloc_len &&
-     (4*new_len >= ea->alloc_len || ea->alloc_len == ea->min_alloc_len))
-    return 1;
-  new_alloc_len = new_len * 2;
-  if (new_alloc_len < ea->min_alloc_len)
-    new_alloc_len = ea->min_alloc_len;
-  new_t = caml_stat_resize_noexc(ea->t, new_alloc_len * sizeof(struct tracked));
-  if (new_t == NULL) return 0;
-  ea->t = new_t;
-  ea->alloc_len = new_alloc_len;
-  return 1;
+  CAMLassert(domain->current->suspended);
+
+  size_t frames =
+    caml_get_callstack(Callstack_size(domain->entries.config),
+                       &domain->callstack_buffer,
+                       &domain->callstack_buffer_len,
+                       alloc_idx);
+  value res = caml_alloc(frames, 0);
+  for (size_t i = 0; i < frames; ++i) {
+    Store_field(res, i, Val_backtrace_slot(domain->callstack_buffer[i]));
```

Comment:
```suggestion
    Field(res, i) = Val_backtrace_slot(domain->callstack_buffer[i]));
```

We are writing a Long to a just-allocated block, so this should be safe.